### ISSUE TITLE
Load JSON data into BigQuery as structured data (STRUCT type)

### ISSUE THREAD

**Issue by zestyping** [airbytehq/airbyte#1927]
## Tell us about the problem you're trying to solve
I'd like to the structure of the JSON in my data source (e.g. Google AdWords) to be available in my destination, BigQuery, so that I can run meaningful queries on the data.  However, with Airbyte, I only have two options:
(a) Load the JSON data as a big string in one `_airbyte_data` column, which I can't query
(b) Use Basic Normalization, which doesn't handle hierarchical data

## Describe the solution you’d like
BigQuery is already built to handle hierarchical data, so I'd love to have JSON simply load into BigQuery as structured records.  The usual `bq` command can do this, for example: https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-json#loading_json_data_into_a_new_table

Note that it's necessary to provide a schema.  I would recommend using https://pypi.org/project/bigquery-schema-generator/, which will scan the file and produce a schema.  It's worked great for me so far.  The `–autodetect` option to `bq` won't cut it, because it only scans 100 rows, so if a new field appears in some other row, the entire import will fail.  `bigquery-schema-generator` will scan the whole file and come up with a schema that can accommodate every record.

## Describe the alternative you’ve considered or used
The way I've usually done this is to export the data from my data source into a big JSONL file (newline-delimited JSON), then run `bigquery-schema-generator` over it to produce a JSON schema file.  Then I use the BigQuery console to load the data; it takes JSONL as one of its source formats and accepts a JSON schema, and that works nicely.

In theory, I could work around the problem by exporting all the data from the `_airbyte_data` column and then going through the same procedure of generating a schema and then manually uploading the data.  But it would make a lot more sense for Airbyte to simply load the data in its true, raw form.  Arguably, the "most raw" form is the JSON, not a table of rows that contain JSON serialized to strings.

**Comment by cgardens** [comment]
Thanks for reporting. (Also _love_ the doctor who profile pic).

What you're describing sounds reasonable to me. You'd rather have the object represented in the native type system that BQ provides instead of raw json. We will be reviewing everything around "normalization" in the next month, so I'll make sure that keep this in mind as we do!

I'd like to understand this a little better.

> (a) Load the JSON data as a big string in one _airbyte_data column, which I can't query

BQ does have some [json extraction syntax](https://cloud.google.com/bigquery/docs/reference/standard-sql/json_functions) for dealing with "raw" json. Have you worked with this before and found it isn't enough? To be clear, I totally believe that this is a pain point, but I haven't personally experienced it, so I'd like to hear it from someone who has on what makes it really hard to use.

**Comment by zestyping** [comment]
Hi @cgardens!  Thanks.

> You'd rather have the object represented in the native type system that BQ provides instead of raw json.

Exactly!

> BQ does have some json extraction syntax for dealing with "raw" json. Have you worked with this before and found it isn't enough? To be clear, I totally believe that this is a pain point, but I haven't personally experienced it, so I'd like to hear it from someone who has on what makes it really hard to use.

True, it does have these functions.  I have used them in some queries, and they do work as documented.  There are three big drawbacks that I can see:

* As you can imagine, attempting to refer to fields inside deeply nested structures would involve some very verbose expressions that would be hard to read and impractical to maintain.  You'd always be writing `JSON_EXTRACT_SCALAR(_airbyte_data, "$.first_name")` instead of `first_name`, and so on.  Granted, not exactly impossible, but really difficult to work with.

* You can extract individual fields, but you can't extract a child of the JSON object into its own table or view, because the type is unknown.  For example, if the JSON record is `{"name": "Charles", "phone": "+19995551212", "address": {"street": "123 Main St", "city": "Anytown", "state": "AK", "zip": "99999"}}`, you can't SELECT the `address` out of there and get a view or a table with street, city, state, and zip columns.  You also aren't allowed to SELECT the `address` and let the type depend on whatever is in the JSON. That's because [BigQuery is strongly typed](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types); every query has a fixed type that can be computed from the types of the inputs.  To get a table with street, city, state, and zip columns, you have to construct a query with the type `STRUCT<street STRING, city STRING, state STRING, zip STRING>`.

* BigQuery doesn't use traditional database indexes.  However, you can make queries more efficient using partitions or clusters, which work like coarse indexes.  For example, for time-series data, partitioning by date is natural.  To do this, though, you need the date to be in its own column.  This is pretty much analogous to how, with a relational database, you would have to extract the date into its own column in order to index on it.

I hope that helps!  If we choose AirByte, we'd be looking to get going with it pretty soon, so I'd be interested in helping accelerate (or contributing to) the development of this ability.

**Comment by ChristopheDuong** [comment]
Yes, this is totally something we should do soon!

- We probably need to work on the destination on GCS first
- then we can enable the BQ loading as you mentioned:

As commented here:
https://github.com/airbytehq/airbyte/issues/1479#issuecomment-768382496

**Comment by zestyping** [comment]
Thanks, Christophe!  It's not clear to me how this is related to the GCS
destination.  Could you help me understand—why would this feature depend on
that one?

**Comment by ChristopheDuong** [comment]
As you mentioned here, especially in the second sentence:

> BigQuery is already built to handle hierarchical data, so I'd love to have JSON simply load into BigQuery as structured records. The usual bq command can do this, for example: https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-json#loading_json_data_into_a_new_table

The doc you linked is about loading into BigQuery with the BQ command from GCS (Googe Cloud Storage)

> You can load newline delimited JSON data from Cloud Storage into a new BigQuery table

So if Airbyte is able to write data in newline delimited JSON format in Google Cloud Storage, then it's a matter of appending a BQ load command (or the equivalent in Java/Python/etc) afterward


Of course, you can also load directly from a Local JSON file to BigQuery but the limitations may be problematic...
That's why I am thinking going through GCS first would be the recommended way, what do you think?

> Limitations
> Loading data from a local data source is subject to the following limitations:
>    - Wildcards and comma-separated lists are not supported when you load files from a local data source. Files must be loaded individually.
 >   - When using the Cloud Console, files loaded from a local data source cannot exceed 10 MB. For larger files, load the file from Cloud Storage.

https://cloud.google.com/bigquery/docs/batch-loading-data#loading_data_from_local_files

**Comment by rpedela** [comment]
Loading data into BigQuery from GCS is the fastest and most scalable path. Same goes for exporting, I once exported 1TB in about 5-6 minutes to GCS. Blew my mind.

**Comment by zestyping** [comment]
Ah, you're totally right — that makes sense now.  Our use of Airbyte will depend on this, so I'm highly motivated to help make these features happen if there's a way to do that.

**Comment by cgardens** [comment]
@zestyping we are going to spend some time early next week trying to figure out a plan for what to do here. The file upload problem is somewhat acute, and something we will need to solve for our other cloud databases. We will report back.

We would love to hear more about your use case. We have weekly office hours at 11:30am tomorrow morning ([zoom link](https://us02web.zoom.us/j/84317290190?pwd=S0E5RzZRV2NrTGw1VDM5cFZoV2JCUT09)). Would you be available to join us then. If that time doesn't work, would you be willing to set up a brief call and walk us through your use case?

**Comment by zestyping** [comment]
Yeah, I can join tomorrow.  Thanks for the invitation!


—Ping

**Comment by zestyping** [comment]
11:30 am in what time zone?

**Comment by jrhizor** [comment]
11:30am PST

**Comment by cgardens** [comment]
@zestyping I sent you the wrong time. it's 11am PST (not 11:30). If that earlier time doesn't work for you feel free to just join that zoom link at 11:30 as we originally schedule and I'll still be on. Sorry for the confusion.

**Comment by ChristopheDuong** [comment]
We've handled the general cases of unnesting for all warehouse destinations here: https://github.com/airbytehq/airbyte/pull/2044

But I didn't have time to look at making a specific case for BigQuery to produce struct/record column types instead of pure json string yet

**Comment by ChristopheDuong** [comment]
It doesn't seem like normalization code should generate DBT models that parse a JSON string with BQ then build the STRUCT types for the final table...

The standard way is to go through a CSV/JSON staging file and load into BigQuery while specifying the JSON Schema of the file:
https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-csv#java

For the implementation part, it could use something like this in Java: https://github.com/googleapis/google-cloud-java/issues/2753

**Comment by zestyping** [comment]
Yes, a new BigQuery destination that passes the data in through GCS (instead of normalization) would be ideal! The only reason I'm looking at DBT transformations myself is as a temporary workaround because it seems like the ideal solution may not happen for a long time -- it's been 4 months since this issue was opened and I'm not clear on what's reasonable to expect.

**Comment by ChristopheDuong** [comment]
Yes, we haven't prioritized this issue since February as we've introduced and worked on the general approach to unnest data at that time. (which explodes nested stream in separate tables, even on BigQuery)

But I am now slowly taking a look back again at this issue and in the next 1-2 week, I might be able to write this new destination. 

However, for the approach, I am thinking of uploading from a local JSON file directly to BigQuery without going through GCS first. Since we are batching locally the records anyway, we should be able to avoid the limitations I commented here: https://github.com/airbytehq/airbyte/issues/1927#issuecomment-772408835

Once that iteration is complete, it'll still be possible to extend the ability to a GCS staging area later on if necessary too.

**Comment by cgardens** [comment]
@zestyping,  @ChristopheDuong is working on this this week.

**Comment by zestyping** [comment]
Wonderful!
